[narrow] Better regexp searching

dsedivec opened the pull request:
Three potential improvements here.  Here's how you can test each:

399af4e: Run `dired-narrow-regexp`, type any invalid regexp such as just a backslash (`\`).  Before this commit, you'll immediately get an error and any further input to `dired-narrow-regexp` will have no effect (though it lets you keep typing).  After this commit you can finish your regexp, such as `\(x\)`.

0dce1f7: Before this change, run `dired-narrow-regexp` and try to find a file name using the `^` to anchor the match at the beginning of the file.  For example, `^dired-narrow` will not match anything.  After this change, `^` matches beginning of file names, which I think is probably useful.

368277b04ad0a224f377f83b1fc28d98076e474b: I think this probably affects any dired-narrow command, but I was testing with `dired-narrow-regexp`, so run that, type any character that doesn't appear in any file name (ex. `{` is uncommon), then backspace it, enter any pattern that will match a file, and accept it.  Before this commit, when you exit `dired-narrow-regexp` no files will be shown because that's what was matched with the `{` pattern and the `:dired-narrow` text properties were never removed when you backspaced.  After this commit, backspacing and entering a new pattern will correctly narrow to your new pattern.

I hope it's not too ugly that I submitted these all in one PR, let me know if you want me to split them up, or if you'd like any further explanation or find problems.  Thanks for writing and publishing these "hacks"!

Fuco1 commented on dired-narrow.el:
```diff
@@ -264,7 +264,7 @@ read from minibuffer."
 (defun dired-narrow--regexp-filter (filter)
   (condition-case nil
-      (re-search-forward filter (line-end-position) t)
+      (string-match-p filter (buffer-substring (point) (line-end-position)))
```
I think instead of relying on `point` it would be better to fetch the file name at line, for example with `(dired-utils-get-filename 'no-dir)`.

dsedivec replied:
Done!  Thanks for the suggestion.

Fuco1 commented on dired-narrow.el:
```diff
@@ -174,7 +174,8 @@ when `dired-narrow-exit-when-one-left' and `dired-narrow-enable-blinking' are tr
   "Restore the invisible files of the current buffer."
   (let ((inhibit-read-only t))
-    (remove-text-properties (point-min) (point-max) '(invisible))
+    (remove-text-properties (point-min) (point-max) '(invisible nil
```
I'm not sure what the `nil`s are doing here?  From the docs of the function

```
The third argument PROPERTIES is a property list
whose property names specify the properties to remove.
(The values stored in PROPERTIES are ignored.)
```

I'm on Emacs 25 still, maybe the API was extended in newer Emacs?  Or is this simply a mistake?

dsedivec replied:
Note "property list" not "list of property names".  The [info docs][] actually say (emphasis added):

>The argument props specifies which properties to delete. It should have the form of a property list (see Property Lists): **a list whose elements are property names alternating with corresponding values**. **But only the names matter**—the values that accompany them are ignored. For example, here's how to remove the face property.
>
>     (remove-text-properties start end '(face nil))

[info docs]: https://www.gnu.org/software/emacs/manual/html_node/elisp/Changing-Properties.html

However `remove-text-properties` iterates over that plist, it must not particularly care when you pass it a list with only a single property name and no value.  However, when I added a second property to be deleted, I assumed I would need to follow the format of a plist.

No matter, though: while looking up the documentation here I found `remove-list-of-text-properties`, which just wants a simple list of properties to remove.  I changed my commit to use that instead of `remove-text-properties`.

Fuco1 commented:
@dsedivec This is excellent! Sorry for keeping you wait.  Having one PR is fine so long as the commits are logically separated, which they are here :+1: 

Just some minor details to fix/clarify and it's good to go!

dsedivec commented:
> @dsedivec This is excellent! Sorry for keeping you wait. Having one PR is fine so long as the commits are logically separated, which they are here 👍
> 
> Just some minor details to fix/clarify and it's good to go!

Ha, as you can see by my own turnaround time, your "delay" was *far* from unreasonable. :smile:

I think I have addressed your two comments, and I have updated this PR (rebased, so replaced existing commits not added new commits).  Please let me know if you want any further changes.

Fuco1 commented:
@dsedivec Great, I don't see any more obstacles, so I'm merging.  Thanks!
